### ISSUE TITLE
[Question] Getting Started Documentation for Robot Inventor?

### ISSUE THREAD

**Issue by johnscary-ev3** [pybricks/support#167]
**Question**
Is there documentation for how to get started using Pybricks with the Robot Inventor?

**Context**
Seems like it is being used to do projects by some, however the Docs don't show how to get going.

**Screenshots**
There is a saying that a picture is worth a 1000 words. Screenshots really help to identify and solve problems.

**Comment by laurensvalk** [comment]
Thanks for asking John! You're right, we're quite far along with Pybricks for the Inventor Hub and the Prime Hub. The documentation currently says [coming soon](https://docs.pybricks.com/en/latest/start_inventor.html), because it is still a bit tricky for beginning users. This is because we haven't finished implementing Bluetooth. This means it can't be used with [Pybricks Code](https://www.code.pybricks.com/) just yet. The only way to use it is to install some Python libraries on your computer.

But since you are asking here, this is a good opportunity to show the more curious users how to get started already.

## Overview

The installation consists of three steps:
1. Running a script on the hub using the official app. This installs the Pybricks firmware alongside the original.
2. Starting Pybricks by holding the right button when you turn it on.
3. Installing a Python package on your computer, so you can run programs.

## 0. NOTE TO NEW USERS

I'm going to add one note of caution just in case:

*If you've never installed Python packages on your computer before, please wait until [this guide is ready](https://docs.pybricks.com/en/latest/start_inventor.html). If you're worried about how to recover your hub if something goes wrong, please wait as well. Proceed at your own risk.*

## 1: Installing the firmware

1. Make sure you are logged into GitHub.
2. Go to the latest [master firmware builds](https://github.com/pybricks/pybricks-micropython/actions?query=workflow%3ABuild+branch%3Amaster).
3. Click on the latest build with a green checkmark (:heavy_check_mark:)
4. Download the `primehub` firmware zip file. It contains the installers for both hubs: 
![image](https://user-images.githubusercontent.com/12326241/99875368-54834800-2bef-11eb-8732-fa876db3fd8b.png)
   - If you have the SPIKE Prime Hub (yellow), open the `install_pybricks.llsp` using the official app.
   - If you have the Robot Inventor Hub (teal green), open the `install_pybricks.lms` using the official app.
5. Connect using USB and run the script.
   - **NOTE**: This is a large script, so it takes a while to download. You will not see download progress. Just wait a few minutes and it will start.
   - Just wait for the script to complete. The hub lights indicate installation progress.
   - Then it reboots and you will see the Bluetooth light blink. Keep waiting, you're almost done.
   - Finally, the hub reboots and you will see the standard firmware as if nothing happened. Success!

## 2: Running Pybricks

1. Turn off the hub and unplug USB
2. Hold the right button and turn the hub on again. If the status light is blue, you are running Pybricks.
3. To run the regular firmware again, just reboot. 

**NOTE**: For the moment, the battery does not automatically charge while running Pybricks. So, make sure to reboot the hub when you are done using it, so it can charge the battery using the standard firmware.

## 3: Installing the Pybricksdev command line tool to download and run programs.

*Coming soon!*

**Comment by BertLindeman** [comment]
@laurensvalk Maybe for the final getting started text on:

> 1: Installing the firmware
> ...
>        If you have the SPIKE Prime Hub (yellow), open the install_pybricks.llsp using the official app.
>        If you have the Robot Inventor Hub (teal green), open the install_pybricks.lms using the official app.

Maybe trivial, but as I have both hubs, I **started** using the "other" hub.
Add the name of the hubs:

`If you have the SPIKE Prime Hub (yellow), open the install_pybricks.llsp using the official` **Spike** `app.`
`If you have the Robot Inventor Hub (teal green), open the install_pybricks.lms using the official` **Robot Inventor** `app.`

**Comment by laurensvalk** [comment]
Done, thanks @BertLindeman ! Looking forward to hear how it goes.

**Comment by BertLindeman** [comment]
@laurensvalk Sure do too. 
Thanks

**Comment by johnscary-ev3** [comment]
Thanks for all the detailed info on how to get going with Pybricks on the Robot Inventor.
I will give it a try.
Good to hear I can use Visual Studio Code also since I enjoyed using that with the EV3.
Can use Git and GitHub and the nice editor also.

**Comment by johnscary-ev3** [comment]
Got stuck on install of pybricksdev.
Managed to get the pybrickdev package, find and install "poetry" ver 1.1.4,  and get python 3.8,
but getting this error below when doing the "poetry install" step.
The directory C:/Users/John/AppData/Local/pypoetry/ does not seem to exist.
Any ideas?

PS C:\Users\John\pybricksdev> poetry install
Installing dependencies from lock file

Package operations: 70 installs, 0 updates, 0 removals

  • Installing ipython-genutils (0.2.0)

  EnvCommandError

  Command C:\Users\John\pybricksdev\.venv\Scripts\pip.exe install --no-deps file:///C:/Users/John/AppData/Local/pypoetry/Cache/artifacts/e6/8e/a3/8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8/ipython_genutils-0.2.0-py2.py3-none-any.whl errored with the following return code 1, and output:
  Processing c:\users\john\appdata\local\pypoetry\cache\artifacts\e6\8e\a3\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\ipython_genutils-0.2.0-py2.py3-none-any.whl
  ERROR: Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: 'C:\\Users\\John\\AppData\\Local\\pypoetry\\Cache\\artifacts\\e6\\8e\\a3\\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\\ipython_genutils-0.2.0-py2.py3-none-any.whl'



  at ~\.poetry\lib\poetry\utils\env.py:1074 in _run
      1070│                 output = subprocess.check_output(
      1071│                     cmd, stderr=subprocess.STDOUT, **kwargs
      1072│                 )
      1073│         except CalledProcessError as e:
    → 1074│             raise EnvCommandError(e, input=input_)
      1075│
      1076│         return decode(output)
      1077│
      1078│     def execute(self, bin, *args, **kwargs):

PS C:\Users\John\pybricksdev> poetry --version
Poetry version 1.1.4
PS C:\Users\John\pybricksdev>

**Comment by BertLindeman** [comment]
Just to report install on `Linux Mint  20 Ulyana` was successful.
As I do not want to dive in the poetry configuration and want to run programs in different directories made a small BASH script: `runpybricks.sh`. 

```bash
#!/bin/bash -u
#
# run primehub or inventorhub on USB
#

parm1="${1:-}"
parms="${*}"
# use default program if none is given:
if [ "${parms}" = "" ]; then
    parm1="/home/bert/py/pybricks/tools/show_firmware_version.py"
fi

# take the complete path also if the parm is a relative path
parm1="`realpath ${parm1}`"

# ignore the first parm
shift
parm_rest="$*"

set -x

# switch to the install location of the tools
cd /home/bert/Downloads/pybricksdev/pybricksdev
# poetry run pybricksdev run usb "Pybricks Hub" /home/bert/py/pybricks/${parms}
poetry run pybricksdev run usb "Pybricks Hub" ${parm1} ${parm_rest}


```

So I can "`runpybricks.sh`" in any directory.
 
@laurens If (when ;-) ) I find issues while running this way, should I create a separate issue for them or do you want them added here?
Assuming a new issue, now I re-read this.

Thanks for all the fun, It's great. 
Bert

**Comment by laurensvalk** [comment]
Sounds good @BertLindeman! Yes, feel free to create new issues as you find them. 

Issues with the hub/motors/sensors are probably the most relevant, since Pybricksdev is only a temporary development tool.

We will probably keep using the llsp/lms project for the firmware installation, so that’s another one where bug reports are quite welcome. If we can, we’ll possibly run it from Pybricks Code directly so you don’t have to switch apps (and so you don’t need windows or Mac).

**Comment by Alitai12** [comment]
> Got stuck on install of pybricksdev.
> Managed to get the pybrickdev package, find and install "poetry" ver 1.1.4, and get python 3.8,
> but getting this error below when doing the "poetry install" step.
> The directory C:/Users/John/AppData/Local/pypoetry/ does not seem to exist.
> Any ideas?
> 
> PS C:\Users\John\pybricksdev> poetry install
> Installing dependencies from lock file
> 
> Package operations: 70 installs, 0 updates, 0 removals
> 
> • Installing ipython-genutils (0.2.0)
> 
> EnvCommandError
> 
> Command C:\Users\John\pybricksdev.venv\Scripts\pip.exe install --no-deps file:///C:/Users/John/AppData/Local/pypoetry/Cache/artifacts/e6/8e/a3/8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8/ipython_genutils-0.2.0-py2.py3-none-any.whl errored with the following return code 1, and output:
> Processing c:\users\john\appdata\local\pypoetry\cache\artifacts\e6\8e\a3\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\ipython_genutils-0.2.0-py2.py3-none-any.whl
> ERROR: Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: 'C:\Users\John\AppData\Local\pypoetry\Cache\artifacts\e6\8e\a3\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\ipython_genutils-0.2.0-py2.py3-none-any.whl'
> 
> at ~.poetry\lib\poetry\utils\env.py:1074 in _run 1070│ output = subprocess.check_output( 1071│ cmd, stderr=subprocess.STDOUT, **kwargs 1072│ ) 1073│ except CalledProcessError as e: → 1074│ raise EnvCommandError(e, input=input_)
> 1075│
> 1076│ return decode(output)
> 1077│
> 1078│ def execute(self, bin, *args, **kwargs):
> 
> PS C:\Users\John\pybricksdev> poetry --version
> Poetry version 1.1.4
> PS C:\Users\John\pybricksdev>

Do you have Python 3.8.2?

Open Powershell as Admin and install poetry "(Invoke-WebRequest -Uri https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py -UseBasicParsing).Content | python -"

Open a new Powershell and check version "poetry --version"

**Comment by Alitai12** [comment]
> Got stuck on install of pybricksdev.
> Managed to get the pybrickdev package, find and install "poetry" ver 1.1.4, and get python 3.8,
> but getting this error below when doing the "poetry install" step.
> The directory C:/Users/John/AppData/Local/pypoetry/ does not seem to exist.
> Any ideas?
> 
> PS C:\Users\John\pybricksdev> poetry install
> Installing dependencies from lock file
> 
> Package operations: 70 installs, 0 updates, 0 removals
> 
> • Installing ipython-genutils (0.2.0)
> 
> EnvCommandError
> 
> Command C:\Users\John\pybricksdev.venv\Scripts\pip.exe install --no-deps file:///C:/Users/John/AppData/Local/pypoetry/Cache/artifacts/e6/8e/a3/8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8/ipython_genutils-0.2.0-py2.py3-none-any.whl errored with the following return code 1, and output:
> Processing c:\users\john\appdata\local\pypoetry\cache\artifacts\e6\8e\a3\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\ipython_genutils-0.2.0-py2.py3-none-any.whl
> ERROR: Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: 'C:\Users\John\AppData\Local\pypoetry\Cache\artifacts\e6\8e\a3\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\ipython_genutils-0.2.0-py2.py3-none-any.whl'
> 
> at ~.poetry\lib\poetry\utils\env.py:1074 in _run 1070│ output = subprocess.check_output( 1071│ cmd, stderr=subprocess.STDOUT, **kwargs 1072│ ) 1073│ except CalledProcessError as e: → 1074│ raise EnvCommandError(e, input=input_)
> 1075│
> 1076│ return decode(output)
> 1077│
> 1078│ def execute(self, bin, *args, **kwargs):
> 
> PS C:\Users\John\pybricksdev> poetry --version
> Poetry version 1.1.4
> PS C:\Users\John\pybricksdev>

What does "poetry config --list" say?
For me: cache-dir = "C:\\Users\\Alitai\\AppData\\Local\\pypoetry\\Cache"

**Comment by laurensvalk** [comment]
@johnscary-ev3 , could you try removing the `.venv` folder and the `poetry.lock` file and running `poetry install` again?

**Comment by Alitai12** [comment]
I'm stuck with this message (on 2 clients): "Python 3.8.2 and poetry 1.1.4"

`C:\Users\Alitai\Downloads\pybricksdev-master\pybricksdev>poetry install
Creating virtualenv pybricksdev in C:\Users\Alitai\Downloads\pybricksdev-master\.venv
Installing dependencies from lock file

  FileNotFoundError

  [WinError 2] The system cannot find the specified file

  at c:\users\alitai\appdata\local\programs\python\python38\lib\subprocess.py:1307 in _execute_child
      1303│             sys.audit("subprocess.Popen", executable, args, cwd, env)
      1304│
      1305│             # Start the process
      1306│             try:
    → 1307│                 hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
      1308│                                          # no special security
      1309│                                          None, None,
      1310│                                          int(not close_fds),
      1311│                                          creationflags,`

**Comment by laurensvalk** [comment]
That error can happen when Git is not installed. I will update the instructions above. Thanks for reporting!

**Comment by Alitai12** [comment]
To run:
poetry run pybricksdev run usb "Pybricks Hub" C:\Users\YOURUSERNAME\Downloads\pybricksdev-master\demo\longdemo.py

and

Open Folder in VSC:
pybricksdev-master (not pybricksdev)

**Comment by johnscary-ev3** [comment]
@laurensvalk , I removed those files and "poetry install" rebuilt them but then gave the same error. See below.
The directory C:/Users/John/AppData/Local/pypoetry/ which it is looking for does not exist.
Maybe something went wrong on installing poetry?
I have Git installed and using it successfully with Visual Studio Code.
Thanks.

----------------------------------------------------------------------------------------------------------------------------

PS C:\users\john\pybricksdev> rm .venv

Confirm
The item at C:\users\john\pybricksdev\.venv has children and the Recurse parameter was not specified. If you continue, all children will be
 removed with the item. Are you sure you want to continue?
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"): A
PS C:\users\john\pybricksdev> rm poetry.lock
PS C:\users\john\pybricksdev> poetry install
The currently activated Python version 3.9.0 is not supported by the project (~3.8).
Trying to find and use a compatible version.
Using python3.8 (3.8.6)
Creating virtualenv pybricksdev in C:\users\john\pybricksdev\.venv
Updating dependencies
Resolving dependencies...

Writing lock file

Package operations: 70 installs, 0 updates, 0 removals

  • Installing ipython-genutils (0.2.0)

  EnvCommandError

  Command C:\users\john\pybricksdev\.venv\Scripts\pip.exe install --no-deps file:///C:/Users/John/AppData/Local/pypoetry/Cache/artifacts/e6/8e/a3/8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8/ipython_genutils-0.2.0-py2.py3-none-any.whl errored with the following return code 1, and output:
  Processing c:\users\john\appdata\local\pypoetry\cache\artifacts\e6\8e\a3\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\ipython_genutils-0.2.0-py2.py3-none-any.whl
  ERROR: Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: 'C:\\Users\\John\\AppData\\Local\\pypoetry\\Cache\\artifacts\\e6\\8e\\a3\\8f37e14310c0072b3fcc4240490bcb42630aa695d069aee89953ebd9f8\\ipython_genutils-0.2.0-py2.py3-none-any.whl'

  WARNING: You are using pip version 20.2.2; however, version 20.2.4 is available.
  You should consider upgrading via the 'C:\Users\John\pybricksdev\.venv\Scripts\python.exe -m pip install --upgrade pip' command.


  at ~\.poetry\lib\poetry\utils\env.py:1074 in _run
      1070│                 output = subprocess.check_output(
      1071│                     cmd, stderr=subprocess.STDOUT, **kwargs
      1072│                 )
      1073│         except CalledProcessError as e:
    → 1074│             raise EnvCommandError(e, input=input_)
      1075│
      1076│         return decode(output)
      1077│
      1078│     def execute(self, bin, *args, **kwargs):

**Comment by dlech** [comment]
Did you do as the warning suggests and run `C:\Users\John\pybricksdev.venv\Scripts\python.exe -m pip install --upgrade pip`?

You can also skip `poetry` altogether if you just want to install the command line tool:

```powershell
py -3.8 -m pip install --upgrade pip
py -3.8 -m pip install pipx
py -3.8 -m pipx ensurepath
py -3.8 -m pipx install https://github.com/pybricks/pybricksdev/archive/master.zip
```

You may have to go as far as logging out and logging back in for the `ensurepath` to fully take effect - it seems some apps will pick it up after restarting the app while others don't.

Then just run `pybricksdev` in a terminal.

**Comment by Alitai12** [comment]
@johnscary-ev3 it says you need python 3.8.2 and not 3.8.6. Maybe that's the issue?

**Comment by dlech** [comment]
Any Python 3.8.x version should work.

**Comment by johnscary-ev3** [comment]
Hi,

  Thanks for the help.
 I managed to get the Inventor Hub firmware installed and got the pybricksdev command line tool installed using pipx as above.
So I can use command line from inside VSC to run pybricks programs on the hub.
However if I try to use the VSC debug "Run USB" command it fails, looking for 'validators' module. See below.
I assume that this module was supposed to have been installed by poetry which did not work for me for some reason.

  Any way it seems that users (not developers) like me would benefit by a more simple VSC extension that just connected to the Inventor hub and downloaded programs to the hub. Probably a new "launch.json" file to do this using the command line tool?
 I am thinking of something like what we have for the EV3.  VSC is a nice environment and EV3 pybricks users would already be familiar with it. Don't really need to mess with the entire pybricksdev package and installation.

Any comments?


------------------------------------------------------------------------------------------------------------------------------------
"Command line"
PS C:\Users\John\pybricksdev> pybricksdev run usb "Pybricks Hub" demo/shortdemo.py
Connecting to COM6
2020-11-23 15:24:07,042:    INFO: Sending: 50%
2020-11-23 15:24:07,106:    INFO: Sending: 100%

Hello
World!
2020-11-23 15:24:08,340:    INFO: Disconnected by client.
PS C:\Users\John\pybricksdev> 


----------------------------------------------------------------------------------------------------------------------------------------
"Run USB"
PS C:\Users\John\pybricksdev>  cd 'c:\Users\John\pybricksdev'; & 'c:\Users\John\pybricksdev\.venv\scripts\python.exe' 'c:\Users\John\.vscode\extensions\ms-python.python-2020.11.371526539\pythonFiles\lib\python\debugpy\launcher' '59686' '--' '-m' 'pybricksdev' 'run' 'usb' 'Pybricks Hub' 'c:\Users\John\pybricksdev\demo\shortdemo.py' 
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.1776.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.1776.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\John\.vscode\extensions\ms-python.python-2020.11.371526539\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
    cli.main()
  File "c:\Users\John\.vscode\extensions\ms-python.python-2020.11.371526539\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 430, in main       
    run()
  File "c:\Users\John\.vscode\extensions\ms-python.python-2020.11.371526539\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 317, in run_module 
    run_module_as_main(target, alter_argv=True)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.1776.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 194, in _run_module_as_main        
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.1776.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\John\pybricksdev\pybricksdev\__main__.py", line 6, in <module>
    from .cli import main
  File "C:\Users\John\pybricksdev\pybricksdev\cli\__init__.py", line 10, in <module>
    import validators
ModuleNotFoundError: No module named 'validators'

**Comment by dlech** [comment]
`pybricksdev` is all we have for now. It will be quite some time before Pybricks is ready for "regular users" on SPIKE Prime/Robot Inventor. We have lots of work to do to get there first.

In VS Code, I would suggest making a build task (`tasks.json`) rather than a debug configuration (`launch.json`) to run `pybricksdev ...`. Otherwise it tries to debug `pybricksdev`, which is, of course, not what you are trying to do.

**Comment by johnscary-ev3** [comment]
Ok, thanks for all the help and work you do on Pybricks. Really appreciate that.

I managed to create a tasks.json file for VSC as you suggested. See below.
It works great as a VSC "build" task or "regular" task to download and run the current file in the editor.
That should keep me going for now, although I am sure I will have more questions about using Pybricks on Inventor hub ;0)


```
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Run Pybricks",
            "type": "shell",
            "command": "pybricksdev",
            "args": [
                "run",
                "usb",
                "Pybricks Hub",
                "${file}"
            ],
            "problemMatcher": [],
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
    ]
}
```

**Comment by laurensvalk** [comment]
Glad to hear you got it working @johnscary-ev3, and thanks for sharing your task settings too.

We're looking forward to your feedback. Thanks a lot everyone for trying this super early version!

**Comment by laurensvalk** [comment]
> Any way it seems that users (not developers) like me would benefit by a more simple VSC extension that just connected to the Inventor hub and downloaded programs to the hub.

We'll focus on making it work with [Pybricks Code](https://www.code.pybricks.com/) first, but having a proper Visual Studio Code extension would be very nice indeed. This sounds like a good future or [sponsored](https://github.com/sponsors/pybricks) project.

**Comment by laurensvalk** [comment]
This one line reproduces it for that commit: https://github.com/pybricks/pybricks-micropython/commit/47e67ba8f94fc09ec58ca00e5671276c8cf7cee0#diff-36a9b4c1bd5a5882c0c831198fc00e605390a13eb204b3f55e716699ce52a3dfR77

So I wonder if we're just getting close to allocating ever so slightly too much static variables on Move Hub.

**Comment by laurensvalk** [comment]
Hello everyone! Good news - we've just finished an [initial version with Bluetooth](https://github.com/pybricks/pybricks-micropython/suites/1620249528/artifacts/29762510). I've updated the installation instructions above.

This means that you can use it with [Pybricks Code](https://www.code.pybricks.com/) now. Other than the firmware install step, everything works just like the other hubs. Now you don't need complicated Python libraries anymore. All you need is Google Chrome (or Chromium).

USB support is dropped (for now), so this may break the workflow for some of you. Fortunately, you can still use Pybrickdev if you want to. Just replace `pybricksdev run usb` with `pybricksdev run ble`. This is updated in the steps above.

We're not actively talking about this early alpha version on social media just yet. We'd love to get your feedback first.

As stated above, battery charging is not yet implemented. Although the hub will automatically safely shut down on low battery, it's still recommended to reboot into the original firmware when you're not using it so it can charge.

And for @johnscary-ev3, this means you can get remote control now as per #154 .

**Comment by johnscary-ev3** [comment]
Hi,
Have been trying to use the new Bluetooth version,
I think I got latest install version dated 12/7/20 and installed it.
Seemed to go ok but when hub comes up in Pybricks mode the blue status light flashes now, 2 blinks and a pause, repeating.
I tried to connect using Bluetooth via Pybricks Code in Chrome and also using the pybricksdev program but there is no connection found. USB does not work either, but I guess that is expected as it is supposed to be BLE now. Correct?
What am I missing?

I also tried to manually connect to Bluetooth device and my PC finds device called RaLinkLinuxClient which I guess is the RI Hub?
But it asks me for a PIN which I don't know.

Thanks for any help.

**Comment by laurensvalk** [comment]
> Seemed to go ok but when hub comes up in Pybricks mode the blue status light flashes now, 2 blinks and a pause, repeating.

This is correct. It blinks, which means that is waiting for a connection.

Which operating system are you using on your computer?

> I tried to connect using Bluetooth via Pybricks Code in Chrome and also using the pybricksdev program but there is no connection found.

Does your computer have Bluetooth Low Energy (BLE)? If not, you will need a BLE USB dongle, which are fairly low cost.

If you have an Android phone or tablet, you can visit Pybricks Code to see if things are working as expected.

> USB does not work either, but I guess that is expected as it is supposed to be BLE now. 

Correct.

> my PC finds device called RaLinkLinuxClient which I guess is the RI Hub

No, that is most likely something else.

**Comment by BertLindeman** [comment]
[not so important]
On WIN 10 Home current Mindstorms app 10.0.3
Have been busy for days (not 24 hours a day, not by far) to get the  official Mindstorms Robot Inventor app  and the hub connected by USB. The app would not open the connect dialog.
Just now connected via bluetooth and started the install program and left the machine do its trick.
It is so hard not to press any key in the first phase of the PyBricks install. 
The Spike-hub was not so difficult, that just worked...
[/not so important]
Now finally the PyBricks firmware is in place and I can do some fun stuff.
First connect and test seem ok.

Thanks

**Comment by laurensvalk** [comment]
> It is so hard not to press any key in the first phase of the PyBricks install.

Could you elaborate on this? Do you mean the lack of download progress in the app?

> The Spike-hub was not so difficult, that just worked...

What was the difference between the SPIKE and Inventor hub/app?

**Comment by BertLindeman** [comment]
> 
> 
> > It is so hard not to press any key in the first phase of the PyBricks install.
> 
> Could you elaborate on this? Do you mean the lack of download progress in the app?

The time needed to load the program I think. 
After pressing the start button (not sure what the correct name is, in Dutch it is "Spelen") it looks like **nothing** is happening. The "stop" button remains gray until the uploaded program is really started. If I remember correctly 
even the start button does no animation.
So my first reaction would be: I did not press start yet. So I leave the PC after pressing start.

> 
> > The Spike-hub was not so difficult, that just worked...
> 
> What was the difference between the SPIKE and Inventor hub/app?

Not sure what you mean with this, so I try an answer..
The Spike app is current at version 1.3.2.
The Spike hub connected to the Spike hub via USB and the PyBricks install did just fine.
I did not try to update the inventor hub via the Spike App. Was not sure that this would be safe to do.
Updated the firmware of the Inventor hub via the spike app. No problem there 
Main problem I think is connect to Mindstorms app on USB.

**Comment by laurensvalk** [comment]
Right - unfortunately the official app shows no download progress. I imagine this might be added at some point, since we aren't the only ones with big programs.

In the long run, we may be able to add USB support to Pybricks Code so you wouldn't need the official app to install it. And then we could show installation progress.

One of these two probably needs to happen before we can say that Pybricks is ready for SPIKE Prime / MINDSTORMS Robot Inventor. I agree that the installation is a bit akward right now.

> Main problem I think is connect to Mindstorms app on USB.

I see, thanks for clarifying! Sounds like that is not related to our install script then.

**Comment by BertLindeman** [comment]
> > Main problem I think is connect to Mindstorms app on USB.
> 
> I see, thanks for clarifying! Sounds like that is not related to our install script then.

Yes, the app does not open the connect dialog, so the PyBrick program is not in the picture yet.

USB support would sure enhance the "PyBricks firmware install"

**Comment by johnscary-ev3** [comment]
Hi,
After some research I figured out that my PCs do not have BLE, aka Bluetooth 4.0   ;0(
Guess they are too old.
The RI Hub does connect ok to these PCs using Bluetooth, but I guess it is using the older Bluetooth protocols for that. 
Looks like I need to get that BLE USB dongle you mentioned. Any particular one needed or suggested?
Thanks.

**Comment by laurensvalk** [comment]
It's hard to guarantee which one will or will not work, but pretty much any BLE should work.

> The RI Hub does connect ok to these PCs using Bluetooth, but I guess it is using the older Bluetooth protocols for that.

Indeed, SPIKE / RI supports both Bluetooth Low Energy and Bluetooth Classic. I think the default app uses classic. All the other hubs have only Bluetooth Low Energy.

We use Bluetooth Low Energy for that reason, and also because Bluetooth Classic cannot be used with the browser like Pybricks Code does.

**Comment by johnscary-ev3** [comment]
Hi,
I got an EVEO Bluetooth 5.0 dongle and it is working great for Pybricks Code and Visual Studio Code using BLE ;0)
Now I can run my RI "Charlie" around the room while it sends status info back to my PC using a Pybricks program.
Thanks very much for the help.

**Comment by laurensvalk** [comment]
Glad to hear it! Looking forward to seeing what you've made :)

**Comment by laurensvalk** [comment]
Update: we've changed the dual boot button combination. I'll update the instructions above.

From now on, you just need the center button so it's easier to operate with one hand.

Short press = standard firmware
Long press / hold = Pybricks firmware

**Comment by laurensvalk** [comment]
Update: since a recent update of the SPIKE app, the firmware installation procedure used here will no longer work. We will be working on an alternative solution.

**Comment by damodaran-ai** [comment]
Happy new year folks! Was wondering if I could ask for some quick help. 

I followed the instructions to install the firmware on the Inventor hub (Teal). Install seemed to work as per instructions.  After the long press I get to the status light blinking blue as described above (2 short flashes, pause) and the pixels are lit up in a 3x3 grid in the middle of the unit. I am unable to get this to work with pybricks code for some reason as it just sits in scanning. The official app is able to connect to the official firmware on the same Mac mini from 2020. Any ideas on how I may debug this? BLE doesn't seem to be the issue given that the machine is Bluetooth 5.0 spec and supports BLE.

**Comment by laurensvalk** [comment]
Happy new year to you too! Thanks for trying Pybricks.

Since it seems to work fine on Windows but not on OSX, I think I’ll move your comment to a separate issue so we can better help you and further investigate.

**Comment by laurensvalk** [comment]
We have just updated the installation instructions. Due to changes in how the official apps work, there is now a separate `download_pybricks` project and a `install_pybricks` project.

Please let us know if you run into any issues. Confirmation that it works on both hubs would be highly appreciated as well.

**Comment by laurensvalk** [comment]
Since recent updates in the official app, this doesn't seem to work. I've updated the post accordingly.

To make sure things become a bit more stable, we will probably be developing our own installation tool at some point, so please stay tuned. :)

**Comment by BertLindeman** [comment]
**Small thing**

In the instructions above we "have" to unplug the USB cable. Did not see it getting plugged-in ;-) 

> 2: Starting Pybricks
    Turn off the hub and unplug USB

The first update of the official firmware needs USB, but after that I think USB is not **needed** anymore.

**even smaller thing**

Would it be a good idea to have this instruction on the code.pybricks doc web pages?
Each time I want to follow these instructions,  I have to search for then. (Sorry, my memory)  

I would accept the rationale: it is not stable enough to do this

**Comment by laurensvalk** [comment]
> Would it be a good idea to have this instruction on the code.pybricks doc web pages?
> Each time I want to follow these instructions, I have to search for then. (Sorry, my memory)
> 
> I would accept the rationale: it is not stable enough to do this

That's indeed why they're here for now. For now, it's only for hackers who don't mind reading through Github discussions :smile:

We really need to make it a lot easier before we put it on Pybricks Code.

**Comment by BertLindeman** [comment]
You **are** making steps. Thanks.
The split of the download and the install of the firmware **does** make it more clear.

Install upon the spikehub (minor so did not want a new issue for this):
Hub has current LEGO firmware (updated a few minutes ago)

- first phase shows progress on the display until all pixels are lit.
- the hub seems to be shutdown (no button/pixel lighted)
- pressed (and kept for a few seconds) the center button.
- bluetooth light varying in color
- the hub switched off.
- long-press starts pybricks firmware and shows firmware a13

 In the end it works oke,

**Comment by laurensvalk** [comment]
I updated the post above  --- it should now work again on the Inventor Hub as well.

**Comment by BertLindeman** [comment]
Re-did the download / install upon the Spikehub.
Similar as my previous install the primehub did not auto reboot after all pixels were lit at the end of the install.

 Pressed and hold the main button until bluetooth light got green and then blinking different colors.
After that also no auto reboot.
Short press starts LEGO firmware. Long press starts pybricks firmware.
Next the inventor hub.
The same procedure as the spike hub.
No auto reboots but fine dual boot after that.

But strangely the inventorhub reports it's firmware as
```
('primehub', '3.0.0a13', 'v3.0.0a13-5-g9a80bef6 on 2021-01-15')
```

**Comment by laurensvalk** [comment]
They both use the same firmware, so that seems about right :)

**Comment by BertLindeman** [comment]
> They both use the same firmware, so that seems about right :)

Oke. I used to look at the hub name in the firmware to determine what hub to instantiate.
Like so:
```
hw_type = version_tuple[0]

if str(hw_type) == "technichub":
    from pybricks.hubs import TechnicHub
    hub = TechnicHub()
elif str(hw_type) == "cplushub":
    from pybricks.hubs import CPlusHub
    hub = CPlusHub()
elif hw_type == "inventorhub":
    from pybricks.hubs import InventorHub
    hub = InventorHub()
elif hw_type == "primehub":
    from pybricks.hubs import PrimeHub
    hub = PrimeHub()
```
Sorry, side-stepping...



Just re-did the inventor hub now with certainty that the battery is full.
No change.

**Maybe add in the instructions to open the console window.**

I was a bit disappointed that I could not copy the text from the console. 
But I cannot blame you guys for that, pitty :-)  You would have fixed that in no-time.
So screenshot:
![image](https://user-images.githubusercontent.com/8142081/104807177-28460b00-57dd-11eb-98df-20608d0a8817.png)

Two remarks about the environment: 
The Spikehub has been done full battery and without any connected devices.
The Inventor hub also full battery but I did that in a model, so with attached sensors and motors.

![image](https://user-images.githubusercontent.com/8142081/104807294-3b0d0f80-57de-11eb-8032-76e396fa8ce3.png)

**Comment by laurensvalk** [comment]
> I used to look at the hub name in the firmware to determine what hub to instantiate.

And that's still fine. In Pybricks, `InventorHub` is just an alias for `PrimeHub`. Both names are available on both hubs; they do the same.

You can use the color sensor to self-identify which one it is :smile:

**Comment by BertLindeman** [comment]
>  In Pybricks, InventorHub is just an alias for PrimeHub. Both names are available on both hubs; they do the same.

Good to know the alias.
Maybe sometime in the future good for a line in a doc item 

> You can use the color sensor to self-identify which one it is 😄

HA ha ha hi hi . . . . Great idea Laurens, Scan the color of the hub to be sure and stop the program if its the wrong color.

**Comment by laurensvalk** [comment]
The installation link above has been updated to `firmware.v3.0.0b1`.

**Comment by laurensvalk** [comment]
> > In Pybricks, InventorHub is just an alias for PrimeHub. Both names are available on both hubs; they do the same.
> 
> Good to know the alias.
> Maybe sometime in the future good for a line in a doc item

There is now a [single page for both hubs](https://docs.pybricks.com/en/latest/hubs/primehub.html) to make things simpler.

**Comment by BertLindeman** [comment]
Part of my comment from [some time ago in this issue](https://github.com/pybricks/support/issues/167#issuecomment-759007876)

> Install upon the spikehub (minor so did not want a new issue for this):
> Hub has current LEGO firmware (updated a few minutes ago)
> 
> * first phase shows progress on the display until all pixels are lit.
> * the hub seems to be shutdown (no button/pixel lighted)
> * pressed (and kept for a few seconds) the center button.
> * bluetooth light varying in color
> * the hub switched off.
> * long-press starts pybricks firmware and shows firmware a13
> 
> In the end it works oke,

Have re-installed the current pybricks firmware on the spike primehub again today.
If the hub is connected via USB to the spike app the auto reboot **does** work.
That did not work for me if bluetooth connected.

Maybe update the install instructions above from

> 4. Run the download_pybricks project in the app.

to
4. Connect the hub via USB to the app and run the download_pybricks project in the app.

That also validates the "unplug USB" later in the instructions.
USB might even be a bit faster? I did not time....

**Comment by ws1088** [comment]
![image](https://user-images.githubusercontent.com/3947543/106371667-d85a5e80-631b-11eb-8c60-0d750578acff.png)
I got this error while pressing the middle button (after i successfully run the program from the IDE)

**Comment by dlech** [comment]
> I got this error while pressing the middle button (after i successfully run the program from the IDE)

See https://github.com/pybricks/support/issues/234

**Comment by BertLindeman** [comment]
Due to #250 

I think it would be good to add in the Getting started doc
**after this**
```
Just wait a few minutes until you hear the usual start sound.
The program stops immediately because it is just data that does nothing by itself.
```
**add this**
```
Open the console log to verify no error occurred while downloading the firmware data
```

That would ensure the console is open in the next step, so progress can be seen there.

**Comment by laurensvalk** [comment]
Thanks --- the installation method is going to be simplified so that this should no longer be an issue.

**Comment by BertLindeman** [comment]
> 
> 
> Thanks --- the installation method is going to be simplified so that this should no longer be an issue.

NOW I see why you want the console log **closed** on the download.

By the way now I did the download a few times: via USB is only a bit faster than via bluetooth
one minute 45 versus just over two minutes.

**Comment by laurensvalk** [comment]
Thanks for the time data. It won't be much faster through Pybricks Code, but we will then be able to start lighting up the LEDs *during* the download already, so it's easier to be patient :)

**Comment by BertLindeman** [comment]
> 
> 
> Thanks for the time data. It won't be much faster through Pybricks Code, but we will then be able to start lighting up the LEDs _during_ the download already, so it's easier to be patient :)

And patient **with a progress indication** is so much better.

**Comment by smr99** [comment]
I got pybricks installed on my Robot Inventor and was able to download and run a "hello world" type program using your chrome-based editor on windows 10.  But my main development is on linux so I wanted to get that running.  Preferably using the CLI tools in pybricksdev with the Visual Studio Code integration.

When I run your suggested example 
> poetry run pybricksdev run ble "Pybricks Hub" demo/shortdemo.py

I get the following output and then it hangs forever -- even after my hub powers off.
> Searching for Pybricks Hub
> Connecting to 38:0B:3C:AA:B6:CE: Pybricks Hub

I am using Python 3.9.  I thought I'd seen some indication that the code is being developed using 3.8 (though I can't find that right now).  Is the python version of issue?

**Comment by giadefa** [comment]
Thanks. So with 3.1 the battery will recharge while the motors are working?
Effectively allowing to run for ever?

**Comment by laurensvalk** [comment]
It would make that aspect equivalent to the stock firmware, yes.

And it depends a bit on how much you use the motors. All six at full power will consume more power than USB can likely deliver. See [this page](https://www.philohome.com/motors/motorcomp.htm) to get an idea of energy consumption.

**Comment by giadefa** [comment]
Nice page. It should be fine.

I'll wait for the battery support then.

**Comment by schmehr** [comment]
Hello everyone!
I'm completely new to this, but I successfully installed Pybricks on my son's RI-Hub using your instructions. Worked fine!
After a long press pybricks starts in connect mode with a flashing blue button, as expected.

When trying to connect Pybricks via Bluetooth to code.pybricks.com, I get this error message:

![Screenshot 2021-04-13 192104](https://user-images.githubusercontent.com/82465533/114594769-ec143080-9c8d-11eb-8ce8-412fcdc71411.png)

I removed any RI/Pybricks-Hub from my Win10-BT settings, but still no success so far.

Maybe anyone has an idea how I can fix this?

Thanks and best regards
Oliver

**Comment by laurensvalk** [comment]
Thanks Oliver, it looks like we need to update the link above to the latest version.

Meanwhile, try installing via beta.pybricks.com. This will get you the latest version.

**Comment by schmehr** [comment]
> Meanwhile, try installing via beta.pybricks.com. This will get you the latest version.

Ok, I should have known this link before. Anyway, it worked fine now.

I found this thread by googling "install pybricks 51515". Is this here still the right place for questions/troubleshooting?

Thanks again!

**Comment by laurensvalk** [comment]
I’m glad to hear you were able to find it that way :)

Yes, this is the right place for now. Pybricks for these hubs is still in beta, which is why the instructions are only here. Eventually we’ll include it on our main website.

**Comment by wes-pro** [comment]
Hello,
I'm getting a grip at Spike/Prime and micropython. I was able to install Pybrics firmware on my device without any problem - thanks for instructions provided here.
I've one question though: I want to use ubluetooth module to test some ideas which require communication with other Lego devices (Lego Mario actually). I see this module is available when I boot into newest standard firmware, however when I'm in Pybrick mode - it is not available.
Are there any reasons behind this? Will it become available in some newer version?

Thank you,
Michal

**Comment by hoppy53** [comment]
Had it installed. My experiences:
1: Although I had the latest Mindstorm software installed on the Hub, it did hang in "preparing Hub". Even after refreshing page.
   After a full reset of the hub in Mindstorms it nicely installed the dual boot.
2:In worked straight away under Windows 10 and under MacOS Big Sur
3: In Linux (Mint) the hub did not want to pair initially.
After clearing the old Pybricks Hub name in the Bluetooth device list it did pair immediately. 

Many thanks for this happening.
A USB connection so I can edit in Thonny would be nice !

**Comment by VegaMansk** [comment]
Error on code.pybricks.com screen . After connection thrue bluetooth. I get the message unexpected error while trying to connect: Connection failed for unknown reason.

I use: Windows 10 Education: Versie: 10.0.19044 Build 19044. Chrome browser Versie 96.04.4664.93 (64 bits). The update of My Inventor (51515) with pybricks (with USB) went ok. Do you have advise for me?

[](url
![Connection screen](https://user-images.githubusercontent.com/25133489/145717156-e80f09df-d768-4401-bf9d-d06f9f658dc4.jpg)
![errorscreen](https://user-images.githubusercontent.com/25133489/145717160-6f4985a2-101c-4160-b915-426c1055b293.jpg)
)

**Comment by dlech** [comment]
Please see #270 for troubleshooting tips and open a new issue with logs/packet capture.

**Comment by VegaMansk** [comment]
I tried to install pybricks again. I have got the following error message:

your hub runs LEGO Firmware version: b'v1.4.01.0000-594ce3d'
Your LEGO firmware version is not (yet) supported.
If you have a newer version that is not listed, please ask for support in https://github.com/pybricks/support/issues/167

Here the complete text:

Installation progress
Something went wrong. Please try again.
from _pybricks.install_pybricks import install; install()
Starting installation script.
Checking installation files.
Files looking good!
Getting firmware info.
Your hub runs LEGO Firmware version: b'v1.4.01.0000-594ce3d'
Your LEGO firmware version is not (yet) supported.
Please install one of the following versions and try again.

    v1.2.01.0101-edc1be7 from SPIKE App v2.0.0.
    v1.2.01.0103-d08b6fe from MINDSTORMS Robot Inventor App v1.3.4 / 10.1.0.
    v1.3.00.0203-71f6a41 from MINDSTORMS Robot Inventor App v2.0.0 / 10.2.0.
    v1.1.01.0002-3e5a121 from SPIKE Prime App v1.3.5.
    v1.2.01.0000-849e6c3 from MINDSTORMS Robot Inventor App v1.3.4 / 10.1.0.

If you have a newer version that is not listed, please ask for support in https://github.com/pybricks/support/issues/167

**Comment by johnscary-ev3** [comment]
Hi guys,
I just got the same error as @VegaMansk shown above.
Will you be working to get the pybricks installer to allow version V1.4,01 ?
Looks like all my RI apps are now 10.3.0 and the one on my iPhone recently forced a hub firmware update to V1.4.01
So it looks like I am stuck now. Any suggestions?
Thanks.

**Comment by laurensvalk** [comment]
We've been working on an updated firmware and installer, but it isn't quite ready yet.

Hopefully we can get the next version out in a few weeks. And if not, we can certainly give this older version a maintenance update to allow V1.4.01, if there's demand for it.

**Comment by johnscary-ev3** [comment]
Thanks for quick reply on this.
Since at least two of us have fallen into this trap, I suspect others will have same problem.
Also, it is not so obvious about how to go back to earlier hub versions as the error note suggests.
So, an installer update to allow new version V1.4.01 to be used would be appreciated.

Seems like having your software and the Lego hub software track each other is tricky business.
I suppose every time Lego does a hub update, you would need to test that your software still works and then add new version to your installer. Is there a large dependence on the Lego hub software that would require a lot of update verification?
If not, maybe just check the hub for a minimum rev level in the installer?

Thanks for the help.

**Comment by laurensvalk** [comment]
That’s right - we verify the version just to make sure everything is working. In this case, v1.4 does not seem to be out yet in my country.

Once the installer is integrated in our main app, most of this complexity will go away, but this is what we have to do for now.

**Comment by johnscary-ev3** [comment]
Not sure if Lego hub updates are linked to App updates but the 10.3 App update on my iPhone shows released 2 weeks ago.

 I could help to check functionality using your software on new hub version if that would help.
 I guess the download would need to work ok first, but I have an application that uses a lot of the features, like motors, lights, all sensors, sounds, remote control, etc.   This is based on the RI "Charlie" model.
 Take care.

**Comment by laurensvalk** [comment]
In the mean time, you can install the official SPIKE firmware, which should still be compatible with our dual-boot installer.

**Comment by SpudGunMan** [comment]
spike is now at LEGO Firmware version: b'v1.3.00.0000-e8c274a', so its too new as well

**Comment by blac1175** [comment]
I also have v1.4.01.0000-594ce3d, as reported by https://install.pybricks.com/, and so the install fails.  For me (in the US, on a modern Android phone) this has been the case at least since I started using my 51515 MINDSTORMS Robot Inventor Lego set on 2021 Dec 25.  I'm also looking forward to trying Pybricks, so looking forward to any updates!

For the record, I've tried rolling back to firmware-backup-v1.2.01.0103-d08b6f, and then istalling pybricks, which works, but when I try to run programs, I get all kinds of weird errors.  I did not bother with this, because I assumed I should just wait until the modern Hub firmware is supported by Pybricks, and then try again.

**Comment by laurensvalk** [comment]
Thanks for your patience everyone! :smile:

**Comment by DanNetor** [comment]
A rollback to the firmware "firmware-backup-v1.2.01.0103-d08b6f". doesn't work for me either. After installing the "old" firmware, I can no longer connect the hub for the dual-boot installation.

The following error:
"Failed to connect. Please wait 60 seconds and try again. Make sure that no other apps are using the hub."

![Screenshot 2022-01-12 095419](https://user-images.githubusercontent.com/93319874/149095295-8c801188-124b-4528-98d6-6050ab19f647.png)
"

**Comment by poldolot** [comment]
> ...
> The following error: "Failed to connect. Please wait 60 seconds and try again. Make sure that no other apps are using the hub."
> ...
I've got the same error on Linux, but on MacOS the installations went through. I've also had v1.4.01.0000-594ce3d firmware that gave me error but after rolling back to v1.3.00.0203-71f6a41 everything went fine and now I can code ;>

**Comment by DanNetor** [comment]
I solved the problem on Win 11 / Edge by deleting the driver (hub) in the device manager. then the hub has reconnected properly.

rollback is always necessary (at least in Germany) because the firmware of the native applications (Lego Mindstorms, Spike) currently "not yet" work for the dual boot.

**Comment by laurensvalk** [comment]
Thanks everyone for sharing your experiences. Good news --- we expect to release the next beta version next week or so.

The installation will be compatible with the latest official firmware, so please stay tuned :)

**Comment by laurensvalk** [comment]
The beta release is now out. Thanks for waiting everyone!

### *Please see https://github.com/pybricks/support/issues/591 for updated instructions*

If you have any questions or concerns regarding the update, please feel free to open a new issue or discussion.

We will be closing this discussion about the alpha version so we don't confuse new users.
